sound/soc/sof/ipc.c

```diff
@@ -366,6 +366,36 @@ void snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
 }
 EXPORT_SYMBOL(snd_sof_ipc_reply);
 
+static void ipc_comp_notification(struct snd_sof_dev *sdev,
+				  struct sof_ipc_cmd_hdr *hdr)
+{
+	u32 msg_type = hdr->cmd & SOF_CMD_TYPE_MASK;
+	struct sof_ipc_ctrl_data *cdata;
+	int ret;
+
+	switch (msg_type) {
+	case SOF_IPC_COMP_GET_VALUE:
+	case SOF_IPC_COMP_GET_DATA:
+		cdata = kzalloc(hdr->size, GFP_KERNEL);
+		if (!cdata)
+			return;
+
+		/* read back full message */
+		ret = snd_sof_ipc_msg_data(sdev, NULL, cdata, hdr->size);
+		if (ret < 0) {
+			dev_err(sdev->dev, "failed to read component event: %d\n", ret);
```

Comment:
if a component can have more that one control, how do we know which of the controls received a notification?
Shouldn't there be something that identifies which control and control type received IPC?

Reply:
We do that, we look up the swidget with the matching comp_id, then we identify the matching kcontrol based on the type and index, but it is done in the snd_sof_control_notify()
Here we just fetch the notification message.